src/common/btree.c

```diff
@@ -4640,3 +4683,183 @@ dbtree_overhead_get(int alloc_overhead, unsigned int tclass, uint64_t otype,
 	return 0;
 }
 
+#define CK_BTREE_NODE_FMT             "Node (off=%#lx)... "
+#define CK_BTREE_NODE_MALFORMED_STR   "malformed - "
+#define CK_BTREE_NON_ZERO_PADDING_FMT CK_BTREE_NODE_MALFORMED_STR "tn_pad_32 != 0 (%#" PRIx32 ")"
+#define CK_BTREE_NON_ZERO_GEN_FMT     CK_BTREE_NODE_MALFORMED_STR "tn_gen != 0 (%#" PRIx32 ")"
+
+/**
+ * Validate the integrity of the btree node.
+ *
+ * \param[in] nd	Node to check.
+ * \param[in] nd_off	Node's offset.
+ * \param[in] ck	Checker.
+ *
+ * \retval DER_SUCCESS	The node is correct.
+ * \retval -DER_NOTYPE	The node is malformed.
+ */
+static int
+btr_node_check(struct btr_node *nd, umem_off_t nd_off, btr_report_fn_t report_fn, void *report_arg,
+	       bool error_on_non_zero_padding)
+{
+	uint16_t unknown_flags;
+
+	D_ASSERT(report_fn != NULL);
+
+	unknown_flags = nd->tn_flags & ~(BTR_NODE_LEAF | BTR_NODE_ROOT);
+	if (unknown_flags != 0) {
+		report_fn(report_arg, BTR_REPORT_ERROR,
+			  CK_BTREE_NODE_MALFORMED_STR "unknown flags (%#" PRIx16 ")",
+			  unknown_flags);
+		return -DER_NOTYPE;
+	}
+
+	if (nd->tn_pad_32 != 0) {
+		if (error_on_non_zero_padding) {
+			report_fn(report_arg, BTR_REPORT_ERROR,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_PADDING_FMT, nd_off,
+				  nd->tn_pad_32);
+			return -DER_NOTYPE;
+		} else {
+			report_fn(report_arg, BTR_REPORT_WARNING,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_PADDING_FMT, nd_off,
+				  nd->tn_pad_32);
+		}
+	}
+
+	if (nd->tn_gen != 0) {
+		if (error_on_non_zero_padding) {
+			report_fn(report_arg, BTR_REPORT_ERROR,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_GEN_FMT, nd_off, nd->tn_gen);
+			return -DER_NOTYPE;
+		} else {
+			report_fn(report_arg, BTR_REPORT_WARNING,
+				  CK_BTREE_NODE_FMT CK_BTREE_NON_ZERO_GEN_FMT, nd_off, nd->tn_gen);
+		}
+	}
+
+	report_fn(report_arg, BTR_REPORT_MSG, CK_BTREE_NODE_FMT OK_STR, nd_off);
+
+	return DER_SUCCESS;
+}
+
+/**
+ * \struct node_info
+ *
+ * List of node offsets.
+ */
+struct node_info {
+	d_list_t   link;
+	umem_off_t nd_off;
+};
+
+/**
+ * Validate the integrity of a btree.
+ *
+ * \param[in] tcx		Btree context.
+ * \param[in] ck		Checker.
+ *
+ * \retval DER_SUCCESS		The tree is correct.
+ * \retval -DER_NOTYPE		The tree is malformed.
+ * \retval -DER_NONEXIST	The tree is malformed.
+ * \retval -DER_*		Possibly other errors.
+ */
+static int
+btr_nodes_check(struct btr_context *tcx, btr_report_fn_t report_fn, void *report_arg,
+		bool error_on_non_zero_padding)
```

Comment:
I'm wondering if there is a chance we want to pass in a check() callback in the future, or we always only check btree sanity